From 1c6342fded339260591f3e33ea53a3d6ec02c55b Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Fri, 29 Aug 2014 19:01:24 -0400
Subject: [net] bridge: Fix incorrect judgment of promisc

Message-id: <1409338885-30414-13-git-send-email-vyasevic@redhat.com>
Patchwork-id: 89907
O-Subject: [RHEL7 PATCH 12/13] bridge: Fix incorrect judgment of promisc
Bugzilla: 896669
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

From: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>

br_manage_promisc() incorrectly expects br_auto_port() to return only 0
or 1, while it actually returns flags, i.e., a subset of BR_AUTO_MASK.

Signed-off-by: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit e0a47d1f781665d2aa126b804572314d128a0070)
Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/br_if.c b/net/bridge/br_if.c
index b6396d3..3109b68 100644
--- a/net/bridge/br_if.c
+++ b/net/bridge/br_if.c
@@ -153,7 +153,8 @@ void br_manage_promisc(struct net_bridge *br)
     * This lets us disable promiscuous mode and write
     * this config to hw.
     */
-   if (br->auto_cnt <= br_auto_port(p))
+   if (br->auto_cnt == 0 ||
+       (br->auto_cnt == 1 && br_auto_port(p)))
     br_port_clear_promisc(p);
    else
     br_port_set_promisc(p);
-- 
1.7.1